src/lib/operators/aggregate_hash.cpp

```diff
@@ -215,54 +215,80 @@ void AggregateHash::_aggregate() {
       const auto column_id = _groupby_column_ids.at(group_column_index);
       const auto data_type = input_table->column_data_type(column_id);
 
+      const auto chunk_count = input_table->chunk_count();
+
       resolve_data_type(data_type, [&](auto type) {
         using ColumnDataType = typename decltype(type)::type;
 
-        /*
-        Store unique IDs for equal values in the groupby column (similar to dictionary encoding).
-        The ID 0 is reserved for NULL values. The combined IDs build an AggregateKey for each row.
-        */
-
-        // This time, we have no idea how much space we need, so we take some memory and then rely on the automatic
-        // resizing. The size is quite random, but since single memory allocations do not cost too much, we rather
-        // allocate a bit too much.
-        auto temp_buffer = boost::container::pmr::monotonic_buffer_resource(1'000'000);
-        auto allocator = PolymorphicAllocator<std::pair<const ColumnDataType, AggregateKeyEntry>>{&temp_buffer};
-
-        auto id_map = std::unordered_map<ColumnDataType, AggregateKeyEntry, std::hash<ColumnDataType>, std::equal_to<>,
-                                         decltype(allocator)>(allocator);
-        AggregateKeyEntry id_counter = 1u;
-
-        const auto chunk_count = input_table->chunk_count();
-        for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
-          const auto chunk_in = input_table->get_chunk(chunk_id);
-          if (!chunk_in) continue;
-
-          const auto base_segment = chunk_in->get_segment(column_id);
-
-          ChunkOffset chunk_offset{0};
-          segment_iterate<ColumnDataType>(*base_segment, [&](const auto& position) {
-            if (position.is_null()) {
+        if constexpr (std::is_same_v<ColumnDataType, int32_t>) {
+          // For values smaller than AggregateKeyEntry, we can use the value itself as an AggregateKeyEntry. We cannot
```

Comment:
AggregateKeyEntry is the typ of the aggregate key, what does "smaller than" mean here?

Reply:
And where do we ensure that is still fits even with the special NULL handling?

Reply:
> AggregateKeyEntry is the typ of the aggregate key

The type of the aggregate key is AggregateKey, the template argument to _aggregate:

```
  switch (_groupby_column_ids.size()) {
    case 0:
    case 1:
      // No need for a complex data structure if we only have one entry
      _aggregate<AggregateKeyEntry>();
      break;
    case 2:
      // We need to explicitly list all array sizes that we want to support
      _aggregate<std::array<AggregateKeyEntry, 2>>();
      break;
    default:
      PerformanceWarning("No std::array implementation initialized - falling back to vector");
      _aggregate<std::vector<AggregateKeyEntry>>();
      break;
  }
```

> what does "smaller than" mean here?

Would "with a type smaller than" help?

> And where do we ensure that is still fits even with the special NULL handling?

AggregateKeyEntry is 64-bit and we only handle 32-bit here.

Reply:
> Would "with a type smaller than" help?.

Yes. And the same for "values with the same size" ... "types with the same size (i.e., 64bit)" ???

Reply:
- s/values with the same size/types with the same size/
- this way, we avoid building the id map

Reply:
Ack.